onnxruntime/core/common/threadpool.cc

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/platform/threadpool.h"
+
+#ifdef USE_EIGEN_THREADPOOL
+#if defined(_MSC_VER)
+#pragma warning(disable : 4267)
+#endif
+#include <unsupported/Eigen/CXX11/ThreadPool>
+#else
+#include "task_thread_pool.h"
+#endif
+
+namespace onnxruntime {
+
+namespace concurrency {
+#ifdef USE_EIGEN_THREADPOOL
+
+// TODO: This is temporarily taken from Eigen until we upgrade its version.
+// Barrier is an object that allows one or more threads to wait until
+// Notify has been called a specified number of times.
+class Barrier {
+ public:
+  Barrier(unsigned int count) : state_(count << 1), notified_(false) {
+    eigen_assert(((count << 1) >> 1) == count);
+  }
+  ~Barrier() {
+    eigen_assert((state_ >> 1) == 0);
+  }
+
+  void Notify() {
+    unsigned int v = state_.fetch_sub(2, std::memory_order_acq_rel) - 2;
+    if (v != 1) {
+      eigen_assert(((v + 2) & ~1) != 0);
+      return;  // either count has not dropped to 0, or waiter is not waiting
+    }
+    std::unique_lock<std::mutex> l(mu_);
+    eigen_assert(!notified_);
+    notified_ = true;
+    cv_.notify_all();
```

Comment:
>cv_.notify_all(); [](start = 4, length = 17)

nit: consider calling notify outside the lock

Reply:
Though I agree, this is Eigen code that will get absorb into the Eigen dependency when I update the version.